File: languages/java/Sort64.java
import java.util.Random;
import java.lang.Math;

public class Sort64 {

  public static
  void random_shuffle(long[] array, int first, int last) {
    Random rng = new Random();
    for (int i = first + 1; i < last; ++i) {
      int randomPlace =  Math.abs(rng.nextInt()) % ((i - first) + 1);
      long tmp = array[randomPlace];
      array[randomPlace] = array[i];
      array[i] = tmp;
    }
  }

  public static
  void iota(long[] array) {
    for (int i = 0; i < array.length; ++i) {
      array[i] = (long)i;
    }
  }

  public static
  void main(String[] args) {
    int min_size = 8;
    int max_size = 16 * 1024 * 1024;

    System.out.printf("%12s %6s %6s\n", "size", "time", "log2");

    long[] data = new long[max_size];
    long[] buffer = new long[max_size];
    iota(data);
    random_shuffle(data, 0, data.length);

    int lg = 3;
    for (int array_size = min_size; array_size <= max_size; array_size *= 2) {
      long time = timeSort(data, buffer, array_size);
      System.out.printf("%12d %6d %6.1f\n", array_size, time / max_size, (double) time / (max_size * lg));
      ++lg;
    }
  }

  static
  long timeSort(long[] data, long[] buffer, int size) {
    int first = 0;
    long startTime = System.nanoTime();
    while (first <= data.length - size) {
      System.arraycopy(data, first, buffer, 0, size);
      Quicksort64.quicksort(buffer, 0, size);
      first += size;
    }
    return System.nanoTime() - startTime;
  }
}

File: languages/cpp/sort64.cpp
#include <cstdio>
#include <cstddef>
#include <cinttypes>
#include <chrono>
#include <vector>
#include <numeric>
#include <algorithm>

#include "quicksort64.h"

using namespace std;
using namespace std::chrono;

uint64_t nano_time() {
  auto now = high_resolution_clock::now().time_since_epoch();
  return duration_cast<nanoseconds>(now).count();
}

uint64_t time_sort(int64_t* data, int64_t* data_end, int64_t* buffer, size_t size) {
  auto startTime = nano_time();
  while (size <= data_end - data) {
    copy(data, data + size, buffer);
    quicksort(buffer, buffer + size);
    data += size;
  }
  return nano_time() - startTime;
}

int main() {
  size_t min_size(8);
  size_t max_size(16 * 1024 * 1024);

  printf("           %12s %6s %6s\n", "size", "time", "log2");

  vector<int64_t> data(max_size);
  vector<int64_t> buffer(max_size);
  int64_t* data_begin = &*begin(data);
  int64_t* data_end = data_begin + max_size;
  int64_t* buffer_begin = &*begin(buffer);

  iota(data_begin, data_end, 0);
  random_shuffle(data_begin, data_end);

  int lg = 3;
  for (size_t array_size = min_size; array_size <= max_size; array_size *= 2) {
    uint64_t time = time_sort(data_begin, data_end, buffer_begin, array_size);
    printf("     [C++] %12zd %6" PRIu64 " %6.1f\n", array_size, time / max_size, (double)time / (max_size * lg));
    ++lg;
  }
}
